dkg-gadget/src/keygen_manager/mod.rs

```diff
@@ -0,0 +1,274 @@
+use dkg_primitives::types::{DKGError, SignedDKGMessage};
+use dkg_runtime_primitives::{
+	crypto::{AuthorityId, Public},
+	keccak_256, DKGApi, MaxAuthorities, MaxProposalLength, SessionId, GENESIS_AUTHORITY_SET_ID,
+};
+use sc_client_api::Backend;
+use sp_arithmetic::traits::SaturatedConversion;
+use sp_runtime::traits::{Block, Header, NumberFor};
+use std::{
+	marker::PhantomData,
+	pin::Pin,
+	sync::{
+		atomic::{AtomicUsize, Ordering},
+		Arc,
+	},
+};
+
+pub struct KeygenManager<B: Block, BE, C, GE> {
+	// governs the workload for each node
+	work_manager: WorkManager<B>,
+	active_keygen_retry_id: Arc<AtomicUsize>,
+	keygen_state: Arc<Atomic<KeygenState>>,
+	_pd: PhantomData<(B, BE, C, GE)>,
+}
+
+impl<B: Block, BE, C, GE> Clone for KeygenManager<B, BE, C, GE> {
+	fn clone(&self) -> Self {
+		Self {
+			work_manager: self.work_manager.clone(),
+			_pd: self._pd,
+			active_keygen_retry_id: self.active_keygen_retry_id.clone(),
+			keygen_state: self.keygen_state.clone(),
+		}
+	}
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+enum KeygenState {
+	Uninitialized,
+	RunningKeygen,
+	RunningGenesisKeygen,
+	GenesisKeygenCompleted,
+	// session_completed denotes the session that executed the keygen, NOT
+	// the generated DKG public key for the next session
+	KeygenCompleted { session_completed: u64 },
+}
+
+// only 1 task at a time may run for keygen
+const MAX_RUNNING_TASKS: usize = 1;
+
+impl<B, BE, C, GE> KeygenManager<B, BE, C, GE>
+where
+	B: Block,
+	BE: Backend<B> + Unpin + 'static,
+	GE: GossipEngineIface + 'static,
+	C: Client<B, BE> + 'static,
+	C::Api: DKGApi<B, AuthorityId, NumberFor<B>, MaxProposalLength, MaxAuthorities>,
+{
+	pub fn new(logger: DebugLogger, clock: impl HasLatestHeader<B>) -> Self {
+		Self {
+			work_manager: WorkManager::<B>::new(
+				logger,
+				clock,
+				MAX_RUNNING_TASKS,
+				PollMethod::Manual,
+			),
+			active_keygen_retry_id: Arc::new(AtomicUsize::new(0)),
+			keygen_state: Arc::new(Atomic::new(KeygenState::Uninitialized)),
+			_pd: Default::default(),
+		}
+	}
+
+	pub fn deliver_message(&self, message: SignedDKGMessage<Public>) {
+		self.work_manager.deliver_message(message)
+	}
+
+	pub fn session_id_of_active_keygen(&self, now: NumberFor<B>) -> Option<JobMetadata> {
+		self.work_manager.get_active_session_ids(now).pop()
+	}
+
+	fn state(&self) -> KeygenState {
+		self.keygen_state.load(Ordering::SeqCst)
+	}
+
+	fn set_state(&self, state: KeygenState) {
+		self.keygen_state.store(state, Ordering::SeqCst);
+	}
+
+	pub async fn on_block_finalized(
+		&self,
+		header: &B::Header,
+		dkg_worker: &DKGWorker<B, BE, C, GE>,
+	) {
+		let now_n = header.number().clone();
+		let now: u64 = header.number().saturated_into();
+		let current_protocol = self.session_id_of_active_keygen(now_n);
+		let state = self.state();
+
+		/*
+		   Session 0 (beginning): immediately run genesis
+		   Session 0 (ending): run keygen for session 1
+		   Session 1 (beginning): run keygen for session 2
```

Comment:
This doesn't seem accurate, can you do more investigation, talk with @1xstj. I believe we always are executing the next keygen at the end of the current session. It is tied together with the notion of a session in fact. We only rotate sessions when we successfully generate and sign the next key.